### DISCUSSION TITLE
Debugging VMs on Bao

### DISCUSSION THREAD

**Discussion by maskedarray** [bao-project/bao-hypervisor#125] [General]
Hi, since Bao relocated the VM images as specified by cache coloring, I was wondering what would be the best way to debug VM images (eg, baremetal/FreeRTOS) on Bao?
My platform is ZCU102.
To be able to use JTAG, I am using xsct to build and download code into the board, I have tried using the xsct builtin debugging functionality (xsdb) but it does not seem to work good.

**Comment by maskedarray** [comment]
I guess I was wrongly looking to debug on physical memory, partially because of previous experience on riscv of which certain versions of hardware and GNU toolchain are very bad at handling virual address debugging. 
But for arm it seems to be working seemlessly. 
I have broadly tried following:
1. Connecting and loading everything (PMUFW, FSBL, TF-A, U-Boot and Bao's image) through xsct and using xsct interface for debugging. It works with virtual memory, but in some cases I had difficulty trapping on breakpoints. I did not like the differences from GDB either.
2. Alternatively, GDB can be used, basically hw_server also opens gdbservers for hardware. But it needs initialization of U-Boot and execution of U-Boot bootm go command. (I spent quite some time figuring out why it was only showing one core). 
3. The best method I found, was to put the necessary files in SD card, switch to SD boot and reset and download bao's image again through xsct and use either xsct or gdb interface to debug.

**Comment by josecm** [comment]
@maskedarray happy you found a workaround to your problem. Cache coloring shouldn't cause any issues with debugging bao. 
But I can share with you a couple a small hack I have used to debug Bao on the ZCU when booting from the SD card. Essentially, inserting an infinite loop somewhere in Bao initilization and booting normally from the sd card. Then, just attach the debugger and step the PC by one instruction manually.

I would also maybe suggest using Vitis for debugging as it essentially provides a nice GUI over xsdb.
